### ISSUE TITLE
the accuracy of ulysses attention and f.scaled_dot_product_attention

### ISSUE THREAD

**Issue by zhangvia** [chengzeyi/ParaAttention#19]
我使用了下面这个脚本来测试ulysses和scaled_dot_product_attention的精度，但发现精度差异较大。但用仓库里的单元测试脚本，又可以通过：
```python
from para_attn import ulysses_attn_func,ring_attn_func


def parse_args():
    parser = argparse.ArgumentParser(description='Test unified attention with configurable sequence length')
    parser.add_argument('--seqlen', type=int, default=1024,
                      help='sequence length (default: 1024)')
    parser.add_argument('--use_bwd', action='store_true',
                        help='whether to test backward pass (default: False)')
    parser.add_argument('--sp_ulysses_degree', type=int, default=4,
                      help='sp_ulysses_degree (default: world_size)')
    parser.add_argument('--sp_ring_degree', type=int,default=None,
                        help='sp_ring_degree (default: world_size//sp_ulysses_degree)')
    return parser.parse_args()

def log(msg, a, rank0_only=False):
    world_size = dist.get_world_size()
    rank = dist.get_rank()
    if rank0_only:
        if rank == 0:
            print(
                f"[Rank#0] {msg}: "
                f"max {a.abs().max().item()}, "
                f"mean {a.abs().mean().item()}",
                flush=True,
            )
        return

    for i in range(world_size):
        if i == rank:
            if rank == 0:
                print(f"{msg}:")
            print(
                f"[Rank#{rank}] "
                f"max {a.abs().max().item()}, "
                f"mean {a.abs().mean().item()}",
                flush=True,
            )
        dist.barrier()


if __name__ == "__main__":
    args = parse_args()
    
    torch.random.manual_seed(0)
    dist.init_process_group("nccl")

    rank = dist.get_rank()
    world_size = dist.get_world_size()
    dtype = torch.bfloat16
    device = torch.device(f"cuda:{rank}")

    batch_size = 1
    seqlen = args.seqlen
    nheads = 32
    d = 1280 // 32
    dropout_p = 0
    causal = False
    deterministic = False
    
    use_bwd = args.use_bwd

    assert seqlen % world_size == 0
    assert d % 8 == 0


    # Prepare inputs
    q = torch.randn(
        batch_size, seqlen, nheads, d, device=device, dtype=dtype, requires_grad=True if use_bwd else False
    )
    k = torch.randn(
        batch_size, seqlen, nheads, d, device=device, dtype=dtype, requires_grad=True if use_bwd else False 
    )
    v = torch.randn(
        batch_size, seqlen, nheads, d, device=device, dtype=dtype, requires_grad=True if use_bwd else False
    )
    dout = torch.randn(batch_size, seqlen, nheads, d, device=device, dtype=dtype)

    dist.broadcast(q, src=0)
    dist.broadcast(k, src=0)
    dist.broadcast(v, src=0)
    dist.broadcast(dout, src=0)


    sp_ulysses_degree = args.sp_ulysses_degree if args.sp_ulysses_degree is not None else world_size
    sp_ring_degree = args.sp_ring_degree if args.sp_ring_degree is not None else world_size // sp_ulysses_degree

    print(
        f"rank {rank}, sp_ulysses_degree: {sp_ulysses_degree}, sp_ring_degree: {sp_ring_degree}"
    )

    mesh = dist.init_device_mesh('cuda',(sp_ulysses_degree, sp_ring_degree),mesh_dim_names=('ud','rd'))

    local_q = torch.chunk(q,world_size,dim=1)[rank].detach().clone()
    local_k = torch.chunk(k,world_size,dim=1)[rank].detach().clone()
    local_v = torch.chunk(v,world_size,dim=1)[rank].detach().clone()

    if use_bwd:
        local_q.requires_grad = True
        local_k.requires_grad = True
        local_v.requires_grad = True

    attn_func = ulysses_attn_func

    if rank == 0:
        print("#" * 30)
        print("# ds-ulysses forward:")
        print("#" * 30)

    print(f"before sp attn forward: {local_q.shape} {local_k.shape} {local_v.shape}")

    local_out = attn_func(
        local_q,
        local_k,
        local_v,
        None,
        dropout_p,
        causal,
        mesh=mesh['ud'],
    )

    # extract local dout
    local_dout = torch.chunk(dout,world_size,dim=1)[rank].detach().clone()

    max_memory = torch.cuda.max_memory_allocated(device) / (1024 * 1024)  # Convert to MB
    print(f"[Rank#{rank}] Maximum GPU memory used: {max_memory:.2f} MB")
    torch.cuda.reset_peak_memory_stats(device)  # Reset stats


    if rank == 0:
        print("#" * 30)
        print("# ulysses backward:")
        print("#" * 30)
    
    # usp attn backward
    if use_bwd:
        local_out.backward(local_dout)

    dist.barrier()

    if rank == 0:
        print("#" * 30)
        print("# local forward:")
        print("#" * 30)

    out_ref = torch.nn.functional.scaled_dot_product_attention(
        q,
        k,
        v,
        dropout_p=dropout_p,
        is_causal=causal,
    )

    if rank == 0:
        print("#" * 30)
        print("# local backward:")
        print("#" * 30)

    if use_bwd:
        out_ref.backward(dout)

    dist.barrier()

    # check correctness
    local_out_ref = torch.chunk(out_ref,world_size,dim=1)[rank].detach().clone()

    log("local (rank) out", local_out, rank0_only=True)
    log("out (distributed) - out_ref (non-distributed) diff", local_out_ref - local_out)

    torch.testing.assert_close(local_out, local_out_ref, atol=1e-2, rtol=0)

    if use_bwd:
        local_dq_ref = torch.chunk(q.grad,world_size,dim=1)[rank].detach().clone()

        log("load_dq", local_q.grad)
        log("dq diff", local_dq_ref - local_q.grad)

        local_dk_ref = torch.chunk(k.grad,world_size,dim=1)[rank].detach().clone()
        log("load_dk", local_k.grad)
        log("dk diff", local_dk_ref - local_k.grad)

        local_dv_ref = torch.chunk(v.grad,world_size,dim=1)[rank].detach().clone()
        log("load_dv", local_v.grad)
        log("dv diff", local_dv_ref - local_v.grad)

    if dist.is_initialized():
        dist.destroy_process_group()


```

**Comment by chengzeyi** [comment]
@zhangvia You seems to misunderstand the meaning of the input parameters of `scaled_dot_product_attention`. The Q, K, and V should be in the `BHSD` format.

**Comment by zhangvia** [comment]
> [@zhangvia](https://github.com/zhangvia) You seem to misunderstand the meaning of the input parameters of `scaled_dot_product_attention`. The Q, K, and V should be in the `BHSD` format.

ah，damn it. thanks for pointing out that. i'm trying to add backward and usp attention to this repository, so i could use it to training flux, hunyuanvideo. Do you think the architectural design of this code project is suitable for extending training functionality? WDYT? 

I originally intended to use the Yunchang library to implement training.but it use progressgroup to handle multi-dimensional parallelism. that is too complicated and not elegant

**Comment by chengzeyi** [comment]
> > [@zhangvia](https://github.com/zhangvia) You seem to misunderstand the meaning of the input parameters of `scaled_dot_product_attention`. The Q, K, and V should be in the `BHSD` format.
> 
> ah，damn it. thanks for pointing out that. i'm trying to add backward and usp attention to this repository, so i could use it to training flux, hunyuanvideo. Do you think the architectural design of this code project is suitable for extending training functionality? WDYT?
> 
> I originally intended to use the Yunchang library to implement training.but it use progressgroup to handle multi-dimensional parallelism. that is too complicated and not elegant

I think backward with Ulysses style attention shoud work out of the box now, though Ring is now supported now. But it is not hard to add support for Ring.

**Comment by zhangvia** [comment]
>though Ring is now supported now. But it is not hard to add support for Ring.

i didn't notice that, did para-attention support unified-attention in yunchang library? i've already done with that. but it will be better if this repository already implement that.

**Comment by chengzeyi** [comment]
> > though Ring is now supported now. But it is not hard to add support for Ring.
> 
> i didn't notice that, did para-attention support unified-attention in yunchang library? i've already done with that. but it will be better if this repository already implement that.

Unified attention is supported. And that’s what we currently use for FLUX inference.
